File: parsers/swift/Sources/Synx/SynxFormatter.swift
// Canonical SYNX text reformatter — sorts keys, normalises indentation,
// preserves leading directives. Mirrors `fmt_canonical` in synx-core/src/lib.rs.
import Foundation

public enum SynxFormatter {

    public static let maxParseDepth = 128

    public static func format(_ text: String) -> String {
        let clamped = SynxParser.clamp(text)
        let lines = clamped.split(separator: "\n", omittingEmptySubsequences: false)
            .map { (sub: Substring) -> String in
                // Strip trailing \r for CRLF inputs.
                if sub.last == "\r" { return String(sub.dropLast()) }
                return String(sub)
            }

        var directives: [String] = []
        var bodyStart = 0
        for (i, line) in lines.enumerated() {
            let t = line.trimmingCharacters(in: .whitespaces)
            if t == "!active" || t == "!lock" || t == "!tool"
                || t == "!schema" || t == "!llm" || t == "#!mode:active" {
                directives.append(t)
                bodyStart = i + 1
            } else if t.isEmpty || t.hasPrefix("#") || t.hasPrefix("//") {
                bodyStart = i + 1
            } else {
                break
            }
        }

        var nodes: [FmtNode] = []
        _ = fmtParse(lines: lines, start: bodyStart, base: 0, depth: 0, nodes: &nodes)
        fmtSort(&nodes)

        var out = ""
        out.reserveCapacity(min(clamped.utf8.count, SynxParser.maxInputBytes) + 64)
        if !directives.isEmpty {
            out.append(directives.joined(separator: "\n"))
            out.append("\n\n")
        }
        fmtEmit(nodes, indentLvl: 0, into: &out)

        // Trim trailing whitespace / newlines, then exactly one newline.
        while let last = out.last, last == "\n" || last == " " || last == "\t" {
            out.removeLast()
        }
        out.append("\n")
        return out
    }

    private struct FmtNode {
        var header: String
        var children: [FmtNode] = []
        var listItems: [String] = []
        var isMultiline: Bool = false
    }

    private static func indentOf(_ line: String) -> Int {
        var i = 0
        let chars = Array(line)
        while i < chars.count && (chars[i] == " " || chars[i] == "\t") { i += 1 }
        return i
    }

    private static func fmtParse(lines: [String], start: Int, base: Int,
                                 depth: Int, nodes: inout [FmtNode]) -> Int {
        if depth > maxParseDepth { return start }
        var i = start
        while i < lines.count {
            let raw = lines[i]
            let t = raw.trimmingCharacters(in: .whitespaces)
            if t.isEmpty { i += 1; continue }
            let ind = indentOf(raw)
            if ind < base { break }
            if ind > base { i += 1; continue }
            if t.hasPrefix("- ") || t.hasPrefix("#") || t.hasPrefix("//") { i += 1; continue }

            var node = FmtNode(header: t)
            node.isMultiline = (t == "|") || (t.hasSuffix(" |"))
            i += 1

            while i < lines.count {
                let cr = lines[i]
                let ct = cr.trimmingCharacters(in: .whitespaces)
                if ct.isEmpty { i += 1; continue }
                let ci = indentOf(cr)
                if ci <= base { break }
                if node.isMultiline || ct.hasPrefix("- ") {
                    node.listItems.append(ct)
                    i += 1
                } else if ct.hasPrefix("#") || ct.hasPrefix("//") {
                    i += 1
                } else {
                    var subs: [FmtNode] = []
                    i = fmtParse(lines: lines, start: i, base: ci, depth: depth + 1, nodes: &subs)
                    node.children.append(contentsOf: subs)
                }
            }
            nodes.append(node)
        }
        return i
    }

    private static func sortKey(_ header: String) -> String {
        var out = ""
        for ch in header {
            if ch == " " || ch == "\t" || ch == "[" || ch == ":" || ch == "(" { break }
            out.append(ch)
        }
        return out.lowercased()
    }

    private static func fmtSort(_ nodes: inout [FmtNode]) {
        nodes.sort { sortKey($0.header) < sortKey($1.header) }
        for i in 0..<nodes.count {
            fmtSort(&nodes[i].children)
        }
    }

    private static func fmtEmit(_ nodes: [FmtNode], indentLvl: Int, into out: inout String) {
        let sp = String(repeating: " ", count: indentLvl)
        let itemSp = String(repeating: " ", count: indentLvl + 2)
        for n in nodes {
            out.append(sp); out.append(n.header); out.append("\n")
            if !n.children.isEmpty {
                fmtEmit(n.children, indentLvl: indentLvl + 2, into: &out)
            }
            for li in n.listItems {
                out.append(itemSp); out.append(li); out.append("\n")
            }
            if indentLvl == 0 && (!n.children.isEmpty || !n.listItems.isEmpty) {
                out.append("\n")
            }
        }
    }
}

File: parsers/java/src/main/java/com/aperturesyndicate/synx/SynxFormatter.java
package com.aperturesyndicate.synx;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/** Canonical SYNX text reformatter. Mirrors {@code fmt_canonical} in {@code synx-core/src/lib.rs}. */
public final class SynxFormatter {

    private SynxFormatter() {}
    public static final int MAX_PARSE_DEPTH = 128;

    public static String format(String text) {
        String clamped = SynxParser.clampText(text);
        String[] lines = clamped.split("\n", -1);
        for (int i = 0; i < lines.length; i++) {
            String l = lines[i];
            if (!l.isEmpty() && l.charAt(l.length() - 1) == '\r') {
                lines[i] = l.substring(0, l.length() - 1);
            }
        }
        List<String> directives = new ArrayList<>();
        int bodyStart = 0;
        for (int i = 0; i < lines.length; i++) {
            String t = lines[i].strip();
            if (t.equals("!active") || t.equals("!lock") || t.equals("!tool")
                || t.equals("!schema") || t.equals("!llm") || t.equals("#!mode:active")) {
                directives.add(t);
                bodyStart = i + 1;
            } else if (t.isEmpty() || t.startsWith("#") || t.startsWith("//")) {
                bodyStart = i + 1;
            } else break;
        }
        List<FmtNode> nodes = new ArrayList<>();
        fmtParse(lines, bodyStart, 0, 0, nodes);
        fmtSort(nodes);

        StringBuilder out = new StringBuilder();
        if (!directives.isEmpty()) {
            out.append(String.join("\n", directives)).append("\n\n");
        }
        fmtEmit(nodes, 0, out);
        // Trim trailing whitespace, ensure single \n.
        int end = out.length();
        while (end > 0) {
            char c = out.charAt(end - 1);
            if (c == '\n' || c == ' ' || c == '\t') end--;
            else break;
        }
        out.setLength(end);
        out.append('\n');
        return out.toString();
    }

    private static final class FmtNode {
        String header;
        List<FmtNode> children = new ArrayList<>();
        List<String> listItems = new ArrayList<>();
        boolean isMultiline;
    }

    private static int indentOf(String line) {
        int i = 0;
        while (i < line.length() && (line.charAt(i) == ' ' || line.charAt(i) == '\t')) i++;
        return i;
    }

    private static int fmtParse(String[] lines, int start, int base, int depth, List<FmtNode> nodes) {
        if (depth > MAX_PARSE_DEPTH) return start;
        int i = start;
        while (i < lines.length) {
            String raw = lines[i];
            String t = raw.strip();
            if (t.isEmpty()) { i++; continue; }
            int ind = indentOf(raw);
            if (ind < base) break;
            if (ind > base) { i++; continue; }
            if (t.startsWith("- ") || t.startsWith("#") || t.startsWith("//")) { i++; continue; }

            FmtNode node = new FmtNode();
            node.header = t;
            node.isMultiline = t.equals("|") || t.endsWith(" |");
            i++;
            while (i < lines.length) {
                String cr = lines[i];
                String ct = cr.strip();
                if (ct.isEmpty()) { i++; continue; }
                int ci = indentOf(cr);
                if (ci <= base) break;
                if (node.isMultiline || ct.startsWith("- ")) {
                    node.listItems.add(ct);
                    i++;
                } else if (ct.startsWith("#") || ct.startsWith("//")) {
                    i++;
                } else {
                    List<FmtNode> subs = new ArrayList<>();
                    i = fmtParse(lines, i, ci, depth + 1, subs);
                    node.children.addAll(subs);
                }
            }
            nodes.add(node);
        }
        return i;
    }

    private static String sortKey(String header) {
        int end = 0;
        while (end < header.length()) {
            char c = header.charAt(end);
            if (c == ' ' || c == '\t' || c == '[' || c == ':' || c == '(') break;
            end++;
        }
        return header.substring(0, end).toLowerCase();
    }

    private static void fmtSort(List<FmtNode> nodes) {
        nodes.sort(Comparator.comparing(n -> sortKey(n.header)));
        for (FmtNode n : nodes) fmtSort(n.children);
    }

    private static void fmtEmit(List<FmtNode> nodes, int indentLvl, StringBuilder out) {
        String sp = " ".repeat(indentLvl);
        String itemSp = " ".repeat(indentLvl + 2);
        for (FmtNode n : nodes) {
            out.append(sp).append(n.header).append('\n');
            if (!n.children.isEmpty()) fmtEmit(n.children, indentLvl + 2, out);
            for (String li : n.listItems) out.append(itemSp).append(li).append('\n');
            if (indentLvl == 0 && (!n.children.isEmpty() || !n.listItems.isEmpty())) {
                out.append('\n');
            }
        }
    }
}
